### PR TITLE
Simplified inference pass

### PR DESCRIPTION
**Pull request by zerlinazl**
Model runs one global forward and backward pass, instead of going between each keyframe pair as in previous version. Propagates entire image dataset at once through a directional pass- if an image has a provided segmentation, that will be used. Otherwise, the previously predicted segmentation will be used as prompt for current prediction.

Hopefully this makes the code easier to understand.

### CODE REVIEW

File: propagate_mask_medsam2.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+        
+        # Merge forward and backward predictions
+        fused_masks = self.merge_bidirectional_masks(video_logits_f, video_logits_b, class_id)
+        return fused_masks
+
+    def merge_bidirectional_masks(self, video_logits_f, video_logits_b, class_id, thresh=0.5):
+        """
+        Merge forward and backward predictions using distance-weighted fusion between keyframe pairs.
+        Inputs:
+            video_logits_f: forward pass logits for all frames
```
Comment (MyYo) [inline propagate_mask_medsam2.py:186]: again, I think video is not the right context here.
I think a better name would be: 
fw_masks_logit
bw_masks_logit

File: propagate_mask_medsam2.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+            video_logits_b: backward pass logits for all frames
+            class_id: class ID being segmented
```
Comment (MyYo) [inline propagate_mask_medsam2.py:188]: Can you explain why class_id is needed here?
The information should be present in fw_masks_logit, bw_masks_logit

File: propagate_mask_medsam2.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+            thresh: threshold for binary mask conversion
+        Returns:
+            fused_masks: dictionary of merged masks for each frame
+        """
+        fused_masks = {}
+        keyframe_indices = import_data_from_roboflow.get_keyframe_indices(class_id)
+
+        # For each pair of consecutive keyframes
+        for i in range(len(keyframe_indices)-1):
```
Comment (MyYo) [inline propagate_mask_medsam2.py:197]: I think you can remove key frames handling here and just loop over all frames from start to finish.

File: propagate_mask_medsam2.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+
+                            fused_masks[t][obj_id] = mask
+                        else:
+                            logit_f = frame_logits_f[obj_id]
+                            fused_logit = logit_f
+                            prob = sigmoid(fused_logit)
+                            mask = (prob > thresh).bool().numpy()
+                            fused_masks[t][obj_id] = mask
+                elif t in video_logits_f:
+                    print(f"Using forward masks for frame {t}")
```
Comment (MyYo) [inline propagate_mask_medsam2.py:233]: Is this a warning? Seems like normal operation.

Avoid printing as it will just spam the consul

File: propagate_mask_medsam2.py [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+                    if t not in fused_masks:
+                        fused_masks[t] = {}
+                    for logit in video_logits_f[t]:
+                        if not isinstance(logit, torch.Tensor):
+                            logit = torch.tensor(logit)
+                        fused_masks[t] = (sigmoid(logit) > thresh).bool().numpy()
+                elif t in video_logits_b:
+                    print(f"Using reverse masks for frame {t}")
```
Comment (MyYo) [inline propagate_mask_medsam2.py:241]: Is this a warning? Seems like normal operation.

Avoid printing as it will just spam the consul

File: propagate_mask_medsam2.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+        Returns:
+            fused_masks (dict): Predicted mask for each image
+            frame_names (list): list of image file names, in order
+        """
+        video_segments_f, video_logits_f = self._propagate_single_direction(class_id)
+        video_segments_b, video_logits_b = self._propagate_single_direction(class_id, reverse=True)
+        
+        # Merge forward and backward predictions
+        fused_masks = self.merge_bidirectional_masks(video_logits_f, video_logits_b, class_id)
+        return fused_masks
```
Comment (MyYo) [inline propagate_mask_medsam2.py:180]: I'm missing a 3D Gaussian smoothing phase on the fused masks

File: propagate_mask_medsam2.py [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+            video_segments: pass predictions
+            video_logits: pass logits
+        """
+        keyframe_indices = import_data_from_roboflow.get_keyframe_indices(class_id)
+        global IMAGE_DATASET_FOLDER_PATH
+        IMAGE_DATASET_FOLDER_PATH = import_data_from_roboflow.get_image_dataset_folder_path()
+
+        # new predictor state for forward pass
+        self.initialize_new_predictor_state()
+        inference_state = self.predictor.init_state(video_path=IMAGE_DATASET_FOLDER_PATH, async_loading_frames=False)
```
Comment (MyYo) [inline propagate_mask_medsam2.py:89]: Since _propagate_single_direction runs continuously on all frames, no need to store predictor in the class memory

File: propagate_mask_medsam2.py [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+            reverse: if True, perform backward pass
+        Returns:
+            video_segments: pass predictions
+            video_logits: pass logits
+        """
+        keyframe_indices = import_data_from_roboflow.get_keyframe_indices(class_id)
+        global IMAGE_DATASET_FOLDER_PATH
+        IMAGE_DATASET_FOLDER_PATH = import_data_from_roboflow.get_image_dataset_folder_path()
+
+        # new predictor state for forward pass
```
Comment (MyYo) [inline propagate_mask_medsam2.py:87]: A more descriptive comment would be
`# Initiate SAM model `

File: propagate_mask_medsam2.py [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+        self.initialize_new_predictor_state()
+        inference_state = self.predictor.init_state(video_path=IMAGE_DATASET_FOLDER_PATH, async_loading_frames=False)
+
+        video_logits = {}
+        video_segments = {}
+        
+        frame_names = import_data_from_roboflow.list_all_images()
```
Comment (MyYo) [inline propagate_mask_medsam2.py:94]: I see you use a lot of if reverse downstream.
A simpler implementation would be here:
if reverse:
   frame_names.reverse()

Then at the end reverse the outputs.
This will save you a lot of if else in this function

File: propagate_mask_medsam2.py [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+        # Initialize predictor
+        self.predictor = build_sam2_video_predictor(
+            config_file= self.config_filepath,
+            ckpt_path= self.checkpoint_filepath,
+            apply_postprocessing=True,
+            # hydra_overrides_extra=hydra_overrides_extra,
+            vos_optimized=  True,
+        )
+
+    def predict_mask(self, predictor, inference_state, frame_idx, prompt_mask, reverse=False):
```
Comment (MyYo) [inline propagate_mask_medsam2.py:34]: I'm still not sure what this function does. We can leave that for the next round of review. But I believe this is an internal function and thus need to start with _. Meaning:
_predict_mask

File: propagate_mask_medsam2.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+                    reverse=reverse
+                )
+            output_masks_binary[i] = predicted_mask
+            output_masks_logit[i] = predicted_logits
+
+        return output_masks_binary, output_masks_logit
+
+    def propagate(self, class_id):
+        """
+        Perform a global forward pass followed by a global backward pass, for specified class.
```
Comment (MyYo) [inline propagate_mask_medsam2.py:129]: Start documentation with
"This function will initialize from sparse segmentation and would propagate the sparse segmentation to all frames."

File: propagate_mask_medsam2.py [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+        Inputs:
+            class_id: COCO class id to segment
+        Returns:
+            output_masks (dict): Predicted mask for each image
+        """
+        output_masks_binary_f, output_masks_logit_f = self._propagate_single_direction(class_id)
```
Comment (MyYo) [inline propagate_mask_medsam2.py:135]: mask_binary_forward
mask_binary_backward

File: propagate_mask_medsam2.py (hunk 2 above)
Comment (MyYo) [reply propagate_mask_medsam2.py:148]: The reason is weighted average. We can remove the weighting for now

File: propagate_mask_medsam2.py [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+
+        return output_masks_binary, output_masks_logit
+
+    def propagate(self, class_id):
+        """
+        Perform a global forward pass followed by a global backward pass, for specified class.
+        Inputs:
+            class_id: COCO class id to segment
+        Returns:
+            output_masks (dict): Predicted mask for each image
```
Comment (MyYo) [inline propagate_mask_medsam2.py:133]: output_masks - is a 3D numpy array (frameNumber, x,y) and not (frameNumber,y,x) - which one?

File: propagate_mask_medsam2.py [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+        """
+        output_masks_binary_f, output_masks_logit_f = self._propagate_single_direction(class_id)
+        output_masks_binary_b, output_masks_logit_b = self._propagate_single_direction(class_id, reverse=True)
+        
+        # Merge forward and backward predictions
+        output_masks = self._merge_bidirectional_masks(output_masks_logit_f, output_masks_logit_b, class_id)
+        return output_masks
```
Comment (MyYo) [inline propagate_mask_medsam2.py:140]: Gaussian smoothing before returning output

File: propagate_mask_medsam2.py [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+                predictor.add_new_mask(
+                    inference_state=inference_state,
+                    frame_idx=i,
+                    obj_id=class_id,
+                    mask=gt_mask,
+                )
+                predicted_mask = gt_mask
+                predicted_logits = gt_mask
+            else:
+                prev_idx = i - 1 if not reverse else i + 1
```
Comment (MyYo) [inline propagate_mask_medsam2.py:109]: prev_idx = process_range[i-1]

File: propagate_mask_medsam2.py [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+                if prev_idx < 0 or prev_idx >= len(frame_names) or output_masks_binary[prev_idx] is None:
```
Comment (MyYo) [inline propagate_mask_medsam2.py:110]: Add a comment to what it is that you are doing here.

File: propagate_mask_medsam2.py [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+
+    def _propagate_single_direction(self, binary_segmentations, reverse=False):
+        """
+        Forward or backward pass for segmentation prediction. Uses ground truth masks for keyframes, otherwise uses previous prediction.
+        Inputs:
+            binary_segmentations: array containing binary segmentation mask for some frames. if no binary segmentation mask for a certain
+                frame, it will be NaN
+            reverse: if True, perform backward pass
+        Returns:
+            output_masks_binary: list of binary mask predictions (ndarray). None if none
```
Comment (MyYo) [inline propagate_mask_medsam2.py:91]: "None if none" is unclear. Can you elaborate?

File: propagate_mask_medsam2.py [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+            output_masks_logit: list of logits for mask predictions (ndarray). None if none
+        """
+        global IMAGE_DATASET_FOLDER_PATH
```
Comment (MyYo) [inline propagate_mask_medsam2.py:94]: It seems that "IMAGE_DATASET_FOLDER_PATH" is only used in this function. Why do we need it as "global"?

File: propagate_mask_medsam2.py [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+        IMAGE_DATASET_FOLDER_PATH = import_data_from_roboflow.get_image_dataset_folder_path()
```
Comment (MyYo) [inline propagate_mask_medsam2.py:95]: add "IMAGE_DATASET_FOLDER_PATH" as input to propagate(..) and _propagate_single_direction(..) to reduce dependency on roboflow in an internal private function.

File: propagate_mask_medsam2.py [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+            output_masks_binary[i] = np.squeeze(predicted_mask)
+            output_masks_logit[i] = np.squeeze(predicted_logits)
+
+        return output_masks_binary, output_masks_logit
+
+    def propagate(self, binary_segmentations, sigma_xy=0.0, sigma_z=0.0, prob_thresh=0.5):
+        """
+        This function will initialize a model from sparse segmentation and propagate the segmentation to all frames.
+        Inputs:
+            binary_segmentations: array containing binary segmentation mask for some frames. if no binary segmentation mask for a certain
```
Comment (MyYo) [inline propagate_mask_medsam2.py:154]: This is not clear.
Chane to:

"binary_segmentations" is an array of masks, with each element corresponding to a different frame in a 3D volume. Elements in the array can either be None, indicating that no mask is provided for the specific frame (meaning we need to propagate to it), or a numpy array of shape (n, m), which specifies whether each pixel in the frame (with dimensions n by m) is inside (1) or outside (0) the mask (key frame).
For example, binary_segmentations = [None, zeros(n,m), None, None, zeros(n,m)] defines key frames in frames 1 and 4.

File: propagate_mask_medsam2.py [hunk 21]
```diff
@@ -0,0 +1,237 @@
...
+                frame, it will be NaN
+            sigma_xy: gaussian smoothing sigma on x and y axes
+            sigma_z: gaussian smoothing sigma on z axis
+            prob_thresh: probability threshold; values above this are set to 1 in the binary mask.
+        Returns:
+            output_masks (array): shape (z,x,y) predicted masks for each image. NaN for slices with no predictions.
```
Comment (MyYo) [inline propagate_mask_medsam2.py:160]: Not clear...

Maybe: output_masks is a 3D numpy matrix shape (z,x,y) that for each pixel defines if it's inside (1) or outside (0) a mask.

File: propagate_mask_medsam2.py [hunk 22]
```diff
@@ -0,0 +1,237 @@
...
+
+from MedSAM2.sam2.build_sam import build_sam2_video_predictor
+import import_data_from_roboflow
+
+class CustomMEDSAM2():
+    def __init__(self, config_filepath, checkpoint_filepath):
+        self.config_filepath = config_filepath
+        self.checkpoint_filepath = checkpoint_filepath
+
+    def initialize_new_predictor_state(self):
```
Comment (MyYo) [inline propagate_mask_medsam2.py:17]: Remove the function...

File: propagate_mask_medsam2.py [hunk 23]
```diff
@@ -0,0 +1,237 @@
...
+            output_masks_logit: list of logits for mask predictions (ndarray). None if none
+        """
+        global IMAGE_DATASET_FOLDER_PATH
+        IMAGE_DATASET_FOLDER_PATH = import_data_from_roboflow.get_image_dataset_folder_path()
+
+        # get valid segmentation indices
+        keyframe_indices = self._get_keyframe_indices_from_sparse_segmentations(binary_segmentations)
+
+        # Initialize SAM model
+        predictor = self.initialize_new_predictor_state()
```
Comment (MyYo) [inline propagate_mask_medsam2.py:101]: Change to: 

predictor = build_sam2_video_predictor(
            config_file= self.config_filepath,
            ckpt_path= self.checkpoint_filepath,
            apply_postprocessing=True,
            # hydra_overrides_extra=hydra_overrides_extra,
            vos_optimized=  True,

File: propagate_mask_medsam2.py [hunk 24]
```diff
@@ -0,0 +1,237 @@
...
+        inference_state = predictor.init_state(video_path=IMAGE_DATASET_FOLDER_PATH, async_loading_frames=False)
+        mask_shape = binary_segmentations[keyframe_indices[0]].shape
+
+        frame_names = import_data_from_roboflow.list_all_images()
```
Comment (MyYo) [inline propagate_mask_medsam2.py:105]: frame_names is not used... you only need the number of frames in the folder. Can't you get it from IMAGE_DATASET_FOLDER_PATH directly?

I suggest changing frame_names --> number_of_frames

**Review by MyYo — CHANGES_REQUESTED** [review]
I think we are getting close to finishing propagate_mask_medsam2.py
Thanks for making those changes!

**Comment by zerlinazl** [comment]
Colab link: https://colab.research.google.com/github/WinetraubLab/3D-segmentation/blob/global-pass/3D-segmentation.ipynb

**Comment by MyYo** [comment]
With regards to the colab. Can you explain this input?
image

It sounds like we need to select either the first "If loading segmentations from Roboflow" or the second "If loading segmentations from folder" options. But in this demo you define both. It's not clear.

**Comment by MyYo** [comment]
I think some sigma is needed especially in z. Can you work with Emilie to propose a figure that would make things smooth?
image
